Change osggid assigner script to do project group setup (SOFTWARE-5619)

williamnswanson opened the pull request:
Add UNIX Cluster Group creation and LDAP Provisioning to the osggid assigner script.
Also renamed the script to "project_group_setup" to fit it's new role.

williamnswanson committed:
Name change and add UNIX cluster group creation

williamnswanson committed:
Add project group LDAP provisioning

brianhlin commented on project_group_setup.py:
```diff
@@ -23,20 +25,22 @@
+  -o outfile           specify output file (default: write to stdout)
+  -t minTimeout        set minimum timeout, in seconds, for API call (default to {MINTIMEOUT})
+  -T maxTimeout        set maximum timeout, in seconds, for API call (default to {MAXTIMEOUT})
+  -h                   display this help text
```
Can you rebase the whitespace-only changes to a separate commit? (last bullet here https://osg-htc.org/technology/software/git-software-development/#avoid-whitespace-noise)

brianhlin commented on project_group_setup.py:
```diff
@@ -23,20 +25,22 @@
 OSPOOL_PROJECT_PREFIX_STR = "Yes-"
 PROJECT_GIDS_START = 200000
 
+
```
Extra whitespace sneaking in

brianhlin commented on project_group_setup.py:
```diff
@@ -58,6 +62,8 @@ class Options:
     endpoint = ENDPOINT
     user = "co_8.william_test"
     osg_co_id = OSG_CO_ID
+    unix_id = CLUSTER_ID
```
```suggestion
    unix_gid = CLUSTER_GID
```

Just in case this script evolves it's better to be clear about which ID

williamnswanson replied:
This value is for specifying which UNIX Cluster in COmanage we want to add UNIX Cluster groups to. For example, the UNIX Cluster for the test registry has an ucid of 10
![image](https://github.com/opensciencegrid/osg-comanage-project-usermap/assets/144060728/4fd432be-f32d-4dac-b55d-b6d366f8aac6)
How's this?:
```suggestion
    ucid = UNIX_CLUSTER_ID
```

brianhlin replied:
Ah, gotcha! I like your suggestion

brianhlin commented on project_group_setup.py:
```diff
@@ -257,11 +280,27 @@ def main(args):
     except urllib.error.HTTPError as e:
         print(e, file=sys.stderr)
         sys.exit(1)
-
```
Bullet #2 https://osg-htc.org/technology/software/git-software-development/#avoid-whitespace-noise

brianhlin commented on project_group_setup.py:
```diff
@@ -238,16 +253,24 @@ def main(args):
+
+            if gid in project_groups:
+                if osggid_index == -1:
+                    project_name = project_id.replace(OSPOOL_PROJECT_PREFIX_STR, "", 1).lower()
```
I know this is what the script did previously but can we get this name from the "osgname" identifier (or whatever it's called) instead?

williamnswanson replied:
I'm not quite sure if you mean the "osg group name" identifier for projects, or the future plan we talked about to transition the creation of project groups to using something like "project_UW_Swanson" as the CO Group name. If the former, this is the part of the code that determines what the value of the "osg group name" identifier should be for a project group. If the later, shouldn't we wait until all of the old-style "Is OSG Project" type projects are set up in LDAP before we change how we determine which groups are projects?

brianhlin replied:
> the future plan we talked about to transition the creation of project groups to using something like "project_UW_Swanson" as the CO Group name

This is a pipe dream for me right now so let's not make any decisions baked on my half-baked design idea :D 

> If the former, this is the part of the code that determines what the value of the "osg group name" identifier should be for a project group.

Ah, right. How about just taking it from the COManage group name instead? Potential brittleness of the string replacement worries me.

williamnswanson replied:
![image](https://github.com/opensciencegrid/osg-comanage-project-usermap/assets/144060728/229734a3-1fe1-4c61-bd9c-1dfc1c804cbf)
Ok, the script now uses the CO group name to create the value of the OSG Group Name identifier.

brianhlin commented on project_group_setup.py:
```diff
@@ -257,11 +280,27 @@ def main(args):
+
+    for project_gid in project_groups:
+        #Provision all project groups
+        call_api2(POST, f"co_provisioning_targets/provision/{options.provision_target}/cogroupid:{project_gid}.json")
```
We should just need to provision projects that we find without cluster groups, right? If I'm reading this correctly, this will make a lot of calls for already provisioned groups

williamnswanson replied:
I wanted to ensure that all non-provisioned projects with UNIX Cluster Groups got provisioned, even if this script wasn't the one to create it, but I couldn't find a way to determine which projects were or weren't provisioned already. If we accept that any projects that get partially set up by hand won't get provisioned we can just provision groups that the script has just made UNIX Cluster Groups for.

brianhlin replied:
> but I couldn't find a way to determine which projects were or weren't provisioned already.

If a group is in LDAP, it's been provisioned so you should be able to do a set difference vs the set of projects.

> If we accept that any projects that get partially set up by hand won't get provisioned we can just provision groups that the script has just made UNIX Cluster Groups for.

With the above, I don't think this quite matters but I think I'm ok with it.

williamnswanson replied:
Added ldap search for unprovisioned groups at commit 085bb8d

brianhlin commented on project_group_setup.py:
```diff
@@ -238,16 +253,24 @@ def main(args):
+                    )
+                    projects_to_assign_identifiers.append(project_data)
+
+                if not gid in clustered_group_ids:
+                    projects_needing_cluster_groups.add(gid)
```
I think this is basically just a `set.difference()`

williamnswanson replied:
Doesn't set.difference() take another set as it's argument? This _could_ be refactored out of the co_group loop as a difference between project_groups and clustered_group_ids, but we need the co_group loop anyway so I thought this would be better than looping over every project group twice. In this next commit I'll likely shuffle things around in a way that will eliminate the project_groups set altogether.

brianhlin replied:
> Doesn't set.difference() take another set as it's argument?

Yup!

> This could be refactored out of the co_group loop as a difference between project_groups and clustered_group_ids, but we need the co_group loop anyway so I thought this would be better than looping over every project group twice.

For me, this is a clarity over efficiency thing. You could get away with stuffing some of these things into functions instead of a `set()` that would provide both the clarity and efficiency.

> In this next commit I'll likely shuffle things around in a way that will eliminate the project_groups set altogether.

Interesting! A data structure containing the list/set/whatever of projects definitely seems like something you'd need for this task. I'm waiting with bated breath to see what you come up with :)

williamnswanson replied:
Moved get_projects_to_setup() functionality into a method for each type of setup needed (which use set differences for projects needing cluster groups and unprovisioned projects)

brianhlin commented on project_group_setup.py:
```diff
@@ -257,11 +280,27 @@ def main(args):
         add_identifier_to_group(gid, type="osggroup", identifier_name=project_name)
```
This isn't part of your changes but I think you can have multiple of the same identifier type so we may have to be careful here.

williamnswanson replied:
The script now only adds an osggroup identifier if one isn't already present.

brianhlin reviewed (changes requested):
In addition to the comments in-line, I've got a few general thoughts:

1. There's a lot of duplication with `group_fixup.py` that should be refactored out into a separate library
2. Lots of your logic is crammed into `main()`. You should split it out into small functions and call them
3. More comments! Some parts of the script are a little unclear and reviewing would be a bit easier with some explanations on the side.
4. It's better to ask for forgiveness than for permission! There are a few spots in your script where your `if` statements should just be `try`/`except`. It'll help with some of the deep nesting and thus readability of the script

williamnswanson committed:
project_groups_setup refactoring

williamnswanson committed:
Add ldap search for unprovisioned projects

Also, move get_projects_to_setup() functionality into several different methods for clarity.

williamnswanson committed:
pip install ldap3 inside github CI action

williamnswanson commented:
How should the library made from refactoring out various shared methods deal with things like variables in options and script default values? Should that information be added as extra arguments to calls to library methods, or should the "library" be more like an IO class where a script creates (for example) a new "APICaller" object which is constructed with all the default var and command line option information?

williamnswanson committed:
Created COManage scripts method library

Created a file to hold methods that are, or will likely be, used by multiple scripts in the osg-comanage-rpoject-usermap repo. Refactored existing scripts to use the new library.

williamnswanson committed:
Fix provisioning a project that needed an osggid

Projects that needed an OSGGID identifier weren't being provisioned if a UNIX cluster group already existed. Added identifier updating after adding missing identifiers and moved adding identifiers to before checking what projects need UNIX cluster groups and provisioning.

williamnswanson committed:
Spelling / wording / whitespace changes

No functional changes

williamnswanson committed:
Add more comments and remove COManage data dict

brianhlin commented on create_project.py:
```diff
@@ -7,18 +7,14 @@
 
 
 SCRIPT = os.path.basename(__file__)
 ENDPOINT = "https://registry.cilogon.org/registry/"
 USER = "co_7.group_fixup"
 OSG_CO_ID = 7
```
I think `ENDPOINT` and `OSG_CO_ID` are useful/reusable enough to go into the library

brianhlin commented on project_group_setup.py:
```diff
@@ -0,0 +1,270 @@
+LDAP_USER = "uid=registry_user,ou=system,o=OSG,o=CO,dc=cilogon,dc=org"
+OSG_CO_ID = 8
+UNIX_CLUSTER_ID = 10
+LDAP_TARGET_ID = 9
```
I think this is all useful stuff to toss into the library as well, perhaps with a `TEST_` prefix?

brianhlin commented on comanage_scripts_utils.py:
```diff
@@ -0,0 +1,222 @@
+            pid = member["Person"]["Id"]
+            path = f"co_provisioning_targets/provision/{prov_id}/copersonid:{pid}.json"
+            responses[pid] = call_api3(POST, path, data, endpoint, authstr)
+    return responses
```
I assume this file is just a move of functions/vars that used to live in `group_fixup.py` so I'm not looking super closely at it

brianhlin commented on project_group_setup.py:
```diff
@@ -0,0 +1,270 @@
+
+def usage(msg=None):
+    if msg:
+        print(msg + "\n", file=sys.stderr)
```
```suggestion
        print(f"{msg}\n", file=sys.stderr)
```

brianhlin commented on project_group_setup.py:
```diff
@@ -0,0 +1,270 @@
+    authstr = None
+    ldap_authtok = None
+    project_gid_startval = PROJECT_GIDS_START
+
```
I would use production for defaults here and override with command line options for your own testing

brianhlin commented on project_group_setup.py:
```diff
@@ -0,0 +1,270 @@
+
+def parse_options(args):
+    try:
+        ops, args = getopt.getopt(args, "u:c:g:l:p:d:f:e:o:h")
```
Heh, the original author was a lover of `getopt` but I think when you start hitting this number of options, `argparse` is a lot nicer. No need to change this but just an FYI for the future when writing CLI interfaces

brianhlin commented on project_group_setup.py:
```diff
@@ -0,0 +1,270 @@
+
+
+def append_if_project(project_groups, group):
+    # If this group has a ospoolproject id, and it starts with "Yes-", it's a project
```
```suggestion
    """If this group has a ospoolproject id, and it starts with "Yes-", it's a project"""
```

May as well use docstrings as they show up nicely in the Python interpreter

brianhlin commented on project_group_setup.py:
```diff
@@ -0,0 +1,270 @@
+    # Get the value of the osggid identifier, if this group has one
+    osggid = utils.identifier_from_list(group["ID_List"], "osggid")
+    # If this group has a osggid, keep a hold of the highest one we've seen so far
+    if osggid is not None:
+        return max(highest_osggid, int(osggid))
+    else:
+        return highest_osggid
```
This should be a try/except

brianhlin commented on project_group_setup.py:
```diff
@@ -0,0 +1,270 @@
+            # Update highest_osggid, if this group has an osggid and it's higher than the current highest osggid.
+            highest_osggid = update_highest_osggid(highest_osggid, group)
+        except TypeError:
+            pass
```
What causes a `TypeError` here? Can we dump a message in case this sort of thing is relevant?

williamnswanson replied:
This used to be an if statement that checked to see if identifier_list was not None before checking to see if it had a Project identifier, now it just passes if the identifier_list is None because an empty list can't contain the identifier we're looking for.

brianhlin commented on project_group_setup.py:
```diff
@@ -0,0 +1,270 @@
+        )
+        return projects_needing_unix_groups
+    except TypeError:
+        return set()
```
Same thing here about the TypeError question and writing a message for logging

brianhlin commented on project_group_setup.py:
```diff
@@ -0,0 +1,270 @@
+            if int(utils.identifier_from_list(project["ID_List"], "osggid")) in project_osggids_to_provision
+        )
+        return projects_to_provision
+    except TypeError:
```
Same thing here about the TypeError question and writing a message for logging

brianhlin commented on project_group_setup.py:
```diff
@@ -0,0 +1,270 @@
+def main(args):
+    parse_options(args)
+
+    # Make all of the nessisary calls to COManage's API for the data we'll need to set up projects.
```
```suggestion
    # Make all of the necessary calls to COManage's API for the data we'll need to set up projects.
```

I find a spellcheck plugin to be very helpful for comments!

brianhlin commented on project_group_setup.py:
```diff
@@ -0,0 +1,270 @@
+    try:
+        main(sys.argv[1:])
+    except OSError as e:
+        print(e, file=sys.stderr)
+        sys.exit(1)
```
I'm pretty sure that this can be shortened to `sys.exit(e)`

brianhlin reviewed (approved):
I'd rename `comanage_scripts_utils.py` -> `comanage_utils.py`. The `scripts` is a bit superfluous.

But otherwise just a few nitpicks! Preapproving

williamnswanson committed:
last fixes for project group setup script pr

williamnswanson committed:
remove unused imports for project-usermap script

williamnswanson committed:
Merge branch 'master' into SOFTWARE-5619.assign-new-project-gids
